torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -5119,3 +5119,88 @@ def test_output_unused_in_loss(self):
+            for _ in range(10):
+                process_group.allreduce(tensors).wait()
+            # Run monitored barrier
+            timeout = timedelta(seconds=0.1)
+            process_group.monitored_barrier(timeout)
+            # All ranks besides 1 call into barrier, rank 0 should report failure
+            # while others report gloo error.
+            failed_rank = 1
+            if self.rank == failed_rank:
+                return
+            if self.rank == 0:
+                with self.assertRaisesRegex(RuntimeError, f"Rank {failed_rank}"):
+                    process_group.monitored_barrier(timeout)
+            else:
+                # It is permissible for other ranks that did not fail to
+                # successfully exit or crash in the monitored barrier, the main
+                # purpose of monitored barrier is to report the rank that hung
+                # on rank 0.
+                try:
+                    process_group.monitored_barrier(timeout)
+                except RuntimeError:
+                    pass
+
+        @require_backend({"gloo", "nccl"})
+        @require_backends_available({"gloo", "nccl"})
+        def test_monitored_barrier_allreduce_hang(self):
+            # tests expected behavior when nonzero rank hangs.
+            if "NCCL_ASYNC_ERROR_HANDLING" in os.environ:
+                del os.environ["NCCL_ASYNC_ERROR_HANDLING"]
+            os.environ["NCCL_BLOCKING_WAIT"] = "1"
+            nccl_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                timeout=timedelta(seconds=2),
+                backend=dist.Backend.NCCL,
+            )
+            gloo_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                backend=dist.Backend.GLOO,
+            )
+            tensors = [
+                torch.ones(10, device=self.rank) * self.rank
+            ]
+            # Let all ranks call allreduce first to set up communicators etc.
+            # Directly simulating error here will run into store issue described
+            # in https://github.com/pytorch/pytorch/issues/54524.
+            nccl_pg.allreduce(tensors).wait()
+            # All ranks besides 0 call into allreduce. This is to simulate a
+            # desync across the world, where some ranks call into
+            # monitored_barrier() and others are stuck in collective comm.
+            if self.rank != 0:
+                with self.assertRaisesRegex(RuntimeError, "Caught collective operation timeout"):
+                    nccl_pg.allreduce(tensors).wait(timedelta(seconds=0.1))
+                return
+
+            # Rank 0 should report timed out rank.
+            monitored_barrier_timeout_seconds = timedelta(seconds=0.1)
+            world_size = int(self.world_size)
+            if world_size == 2:
+                err_regex = "Rank 1"
+            else:
+                err_regex = f"Rank [1-{world_size - 1}]"
```

Comment:
This regex is because for world_size > 2, it is possible for any nonzero rank to timeout out. We hold a unordered_map of the work objects, so the implementation doesn't enforce a specific order of checking the ranks. Will add a follow up change to provide an option to wait for all failed ranks, though.